### PR TITLE
Make multivariate distributions work with `varmat`

### PR DESCRIPTION
**Pull request by bbbales2**
## Summary

This is a pull to make the multivariate distributions work with `varmat`. It will need to wait on #2266 and all the scalar distributions to be converted first.

## Release notes

- Make multivariate distributions work with `varmat`

## Checklist

- [x] Math issue #2101

- [x] Copyright holder: Columbia University

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by bbbales2** [comment]
This is ready after #2266 and #2292 are in.

File: test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp [hunk 1]
```diff
@@ -32,52 +24,125 @@ TEST(mathMixMatFun, traceGenInvQuadForm) {
...
   B << 5, 6, 7, 8;
   Eigen::MatrixXd D(2, 2);
   D << 9, 10, 11, 12;
   stan::test::expect_ad(tols, f, D, A, B);
+  stan::test::expect_ad_matvar(tols, f, D, A, B);
 
   // exception tests
-  // non-square second arg
-  Eigen::MatrixXd a34(3, 4);
-  stan::test::expect_ad(f, c, a34, b);
```
Comment (t4c1) [inline test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp:68]: Why is this deleted?

File: test/unit/math/mix/prob/multi_student_t_test.cpp [hunk 2]
```diff
@@ -1,7 +1,62 @@
...
+#include <test/unit/math/test_ad.hpp>
+
+TEST(ProbDistributionsMultiStudentT, matvar) {
+  auto f
+      = [](const auto& y, const auto& nu, const auto& mu, const auto& sigma) {
+          auto sigma_sym = stan::math::multiply(0.5, sigma + sigma.transpose());
+          return stan::math::multi_student_t_lpdf(y, nu, mu, sigma_sym);
+        };
+
+  auto f_const_y = [](const auto& y) {
```
Comment (t4c1) [inline test/unit/math/mix/prob/multi_student_t_test.cpp:10]: What is the reason for these tests that make one argument constant?

File: test/unit/math/rev/fun/LDLT_factor_test.cpp [hunk 3]
```diff
@@ -86,93 +47,20 @@ TEST(AgradRevMatrix, solve) {
...
-  A << 0, 0, 0, 0;
-  EXPECT_NO_THROW(ldlt_A.compute(A));
-  ASSERT_FALSE(ldlt_A.success());
-
-  // tests on A: [1, 2, 3][2, 3, 4]
-  A.resize(2, 3);
-  A << 1, 2, 2, 3, 3, 4;
-  EXPECT_THROW(ldlt_A.compute(A), std::invalid_argument);
-  ASSERT_FALSE(ldlt_A.success());
+  // EXPECT_MATRIX_EQ(A_double, ldlt_double.matrix());
```
Comment (t4c1) [inline test/unit/math/rev/fun/LDLT_factor_test.cpp:62]: commented code?

File: test/unit/math/test_ad_matvar.hpp [hunk 4]
```diff
@@ -635,6 +635,141 @@ void expect_ad_matvar(const F& f, const EigMat1& x, const EigMat2& y,
...
+ */
+template <typename F, typename EigMat1, typename EigMat2, typename EigMat3,
+          typename EigMat4>
+void expect_ad_matvar(const ad_tolerances& tols, const F& f, const EigMat1& x,
+                      const EigMat2& y, const EigMat3& z, const EigMat4& q) {
+  using stan::math::var;
+  using varmat = stan::math::var_value<Eigen::MatrixXd>;
+
+  // expect_ad_matvar_impl<double, double, double, double>(tols, f, x, y, z, q);
+  // expect_ad_matvar_impl<double, double, double, var>(tols, f, x, y, z, q);
```
Comment (t4c1) [inline test/unit/math/test_ad_matvar.hpp:662]: lots of commented code?

**Review by t4c1 — CHANGES_REQUESTED** [review]
This PR is massive. Next time please please make smaller ones.

File: stan/math/prim/fun/LDLT_factor.hpp [hunk 5]
```diff
@@ -12,120 +12,55 @@ namespace stan {
...
-    vector_t ldltP_diag(ldltP_->vectorD());
-    for (int i = 0; i < ldltP_diag.size(); ++i) {
-      if (ldltP_diag(i) <= 0 || is_nan(ldltP_diag(i))) {
-        return false;
-      }
-    }
-    return true;
-  }
+template <typename T, typename Enable = void>
+class LDLT_factor;
```
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/fun/LDLT_factor.hpp:21]: varmat support was the main thing. The last thing was confusing with how it handled memory and had a lot of unnecessary complexity built in.

File: test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp (hunk 1 above)
Comment (bbbales2) [reply test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp:68]: `a34 + a34.transpose()` cannot be computed (in the symmetrizing step) so this fails before it even hits the function.

File: test/unit/math/mix/prob/multi_student_t_test.cpp (hunk 2 above)
Comment (bbbales2) [reply test/unit/math/mix/prob/multi_student_t_test.cpp:10]: Four argument tests aren't available for `expect_ad`

File: test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp (hunk 1 above)
Comment (bbbales2) [reply test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp:68]: I could add a function back that doesn't do the symmetrize though and it should fail. This should be tested in prim anywho

**Comment by bbbales2** [comment]
> This PR is massive.

Ooof, I didn't merge develop in here after the ldlt pull got merged into develop. It was 22 files not 50.

File: stan/math/fwd/fun/mdivide_left_ldlt.hpp [hunk 6]
```diff
@@ -13,27 +13,27 @@ namespace math {
...
  * @tparam EigMat type of the right-hand side matrix or vector
- *
  * @param A LDLT_factor
  * @param b right-hand side matrix or vector
  * @return x = b A^-1, solution of the linear system.
  * @throws std::domain_error if rows of b don't match the size of A.
  */
-template <int R1, int C1, typename EigMat,
+template <typename T, typename EigMat,
+          require_eigen_vt<std::is_arithmetic, T>* = nullptr,
```
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/fwd/fun/mdivide_left_ldlt.hpp]: I think technically it needs to be like this if the `LDLT_factor` got constructed with something that has a plain type matrix but isn't actually a matrix itself. But since I push everything through `make_ldlt_factor` now I already do the conversion to plain type.

Anyway I'm pretty sure if I changed the signature to `LDLT_factor<Eigen::MatrixXd>` it would work.

File: stan/math/rev/fun/trace_gen_inv_quad_form_ldlt.hpp [hunk 7]
```diff
@@ -17,40 +17,310 @@ namespace math {
...
+      double C_adj = res.adj();
+
+      arena_A.adj()
+          -= C_adj * AsolveB * arena_D.transpose() * AsolveB.transpose();
+      arena_B.adj() += C_adj * AsolveB * (arena_D + arena_D.transpose());
+    });
+
+    return res;
+  } else if (!is_constant<Ta>::value && is_constant<Tb>::value
+             && !is_constant<Td>::value) {
```
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/rev/fun/trace_gen_inv_quad_form_ldlt.hpp]: It is monotonous, but now that I am used to doing it it's not too bad.

Presumably C++17 will make life easier. I'm happy to wait for that.

File: test/unit/math/rev/fun/LDLT_factor_test.cpp (hunk 3 above)
Comment (bbbales2) [reply test/unit/math/rev/fun/LDLT_factor_test.cpp]: Good news! The `.matrix()` function to the prim version of `LDLT_factor` was broken. Uncommenting this revealed a bug. So it is fixed! So your efforts reviewing the giant pull were not for naught.

File: stan/math/prim/fun/LDLT_factor.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/prim/fun/LDLT_factor.hpp]: So now you never need to make a decomposition in forward pass and use it do calculate something in reverse pass?

File: stan/math/rev/fun/trace_gen_inv_quad_form_ldlt.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/rev/fun/trace_gen_inv_quad_form_ldlt.hpp]: This case will require some work regardless of c++17.

File: test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp (hunk 1 above)
Comment (t4c1) [reply test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp]: Yeah, lets add such test back in. Prim is fine.

Comment (bbbales2) [reply test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp]: Alright, the problem with adding a check back is that the new `LDLT_factor` isn't doing `check_symmetric` things. It's leaving that to the functions that use it.

So there's no way to construct the LDLT factor to test `trace_gen_inv_quad_form_ldlt`, and then `LDLT_factor` is used internally only in places where we've already done the symmetry checks.

Would be 10 minutes to add a symmetry check + tests, so pretty easy. You think we should? Otherwise I'll leave it.

File: stan/math/prim/fun/LDLT_factor.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/fun/LDLT_factor.hpp]: Also LDLT_factor was technically done in the previous commit :/

File: test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp (hunk 1 above)
Comment (t4c1) [reply test/unit/math/mix/fun/trace_gen_inv_quad_form_ldlt_test.cpp]: Now that you merged develop in I cant even find this anymore. Im confused. It now looks like nothing was deleted so it is fine I guess.

**Review by t4c1 — APPROVED** [review]
